test/unit/qualityTest.test.js

```diff
@@ -0,0 +1,214 @@
+
+
+const trial = async function(trainSize, testSize, trainer, labelFn) {
+  const trainingOcean = generateOcean(trainSize);
+  const trainingLabels = trainingOcean.map(fish => labelFn(fish));
+
+  const testOcean = generateOcean(testSize);
+  const testData = testOcean.map(fish => fish.knnData);
+
+  for (const fish of trainingOcean) {
+    const label = labelFn(fish);
+    trainer.addTrainingExample(fish.knnData, label);
+  };
+
+  trainer.train();
+
+  for (const fish of testOcean) {
+    fish.result = await trainer.predictFromExample(fish.knnData);
+  }
+
+  return createConfusionMatrix(testOcean, trainSize, labelFn);
+};
+
+const average = function(list) {
+  const sum = list.reduce((prev, curr) => prev + curr);
+  return (1.0 * sum / list.length);
+};
+
+const performTrials = async function({numTrials, trainSize, testSize, createTrainerFn, labelFn}) {
+  const trials = [];
+
+  if (!createTrainerFn) {
+    createTrainerFn = () => {
+      const trainer = new SimpleTrainer();
+      trainer.initializeClassifiersWithoutMobilenet();
+      return trainer;
+    };
+  }
+
+  //console.log(`${numTrials} trials`);
+  for (var i = 0; i < numTrials; i++) {
+    //console.log(`Trial # ${i + 1}`);
+    trials.push(await trial(trainSize, testSize, createTrainerFn(), labelFn));
+  }
+
+  const averagedConfusionMatrix = {};
+  const keys = ['truePos', 'falsePos', 'falseNeg', 'trueNeg', 'precision', 'recall', 'accuracy'];
+
+  for (const key of keys) {
+    //console.log(`${key} ${JSON.stringify(resultsForCurrentNumberOfTrials.map(result => result[key]))}`)
+    const values = trials.map(t => t[key]);
+    const avg = average(values);
+    averagedConfusionMatrix[key] = avg;
+  }
+
+  return averagedConfusionMatrix;
+};
+
+const createConfusionMatrix = function(predictedOcean, numTrained, labelFn) {
+  //const confusionMatrix = {truePos: 0, falsePos: 0, falseNeg: 0, trueNeg: 0}
+  var truePos, falsePos, falseNeg, trueNeg;
+  truePos = falsePos = falseNeg = trueNeg = 0;
+
+  for (const fish of predictedOcean) {
+    const actualLabel = labelFn(fish);
+    const prediction = fish.result.predictedClassId;
+
+    if (prediction && actualLabel) {
+      truePos++;
+    } else if (prediction && !actualLabel) {
+      falsePos++;
+    } else if (!prediction && actualLabel) {
+      falseNeg++;
+    } else {
+      trueNeg++;
+    }
+  };
+
+  const totalSize = truePos + falsePos + falseNeg + trueNeg;
+
+  const precision = truePos / (truePos + falsePos);
+  const recall = truePos / (truePos + falseNeg);
+  const accuracy = (truePos + trueNeg) / totalSize;
+
+  const confusionMatrix = {
+    truePos: truePos,
+    falsePos: falsePos,
+    falseNeg: falseNeg,
+    trueNeg: trueNeg,
+    // If the model didn't predict "yes" for anything in the test data, precision is NaN.
+    // We can assume there was at least one positive in the test data and call this 0 precision.
+    precision: precision || 0,
+    recall: recall,
```

Comment:
Is the comment for precision not possible for recall as well?

Reply:
Not unless we use a small sample, since the denominator for recall is # of actual positives in the sample, so it's almost guaranteed there'll be at least one.